flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsResultPartitionReadSchedulerTest.java

```diff
@@ -0,0 +1,403 @@
+                        ioExecutor,
+                        new HsFileDataIndexImpl(NUM_SUBPARTITIONS),
+                        dataFilePath,
+                        HybridShuffleConfiguration.createConfiguration(
+                                bufferPool.getNumBuffersPerRequest(),
+                                NUM_SUBPARTITIONS,
+                                bufferRequestTimeout));
+
+        TestingHsSubpartitionFileReader.Factory factory =
+                new TestingHsSubpartitionFileReader.Factory();
+        AtomicBoolean prepareForSchedulingFinished = new AtomicBoolean(false);
+        AtomicReference<Throwable> cause = new AtomicReference<>();
+        factory.setPrepareForSchedulingRunnable(() -> prepareForSchedulingFinished.set(true));
+        factory.setFailConsumer((cause::set));
+
+        readScheduler.registerNewSubpartition(0, subpartitionViewOperation, factory);
+
+        ioExecutor.trigger();
+
+        assertThat(prepareForSchedulingFinished).isTrue();
+        assertThat(cause)
+                .hasValueSatisfying(
+                        throwable ->
+                                assertThat(throwable)
+                                        .isInstanceOf(TimeoutException.class)
+                                        .hasMessageContaining("Buffer request timeout"));
+    }
+
+    /**
+     * When {@link SubpartitionFileReader#readBuffers(Queue, BufferRecycler)} throw IOException,
+     * subpartition reader should fail.
+     */
+    @Test
+    void testRunReadBuffersThrowException() throws Exception {
+        TestingHsSubpartitionFileReader.Factory factory =
+                new TestingHsSubpartitionFileReader.Factory();
+        AtomicReference<Throwable> cause = new AtomicReference<>();
+        factory.setFailConsumer((cause::set));
+        factory.setReadBuffersConsumer(
+                (buffers, recycle) -> {
+                    throw new IOException("expected exception.");
+                });
+
+        readScheduler.registerNewSubpartition(0, subpartitionViewOperation, factory);
+
+        ioExecutor.trigger();
+
+        assertThat(cause)
+                .hasValueSatisfying(
+                        throwable ->
+                                assertThat(throwable)
+                                        .isInstanceOf(IOException.class)
+                                        .hasMessageContaining("expected exception."));
+    }
+
+    // ----------------------- test release ---------------------------------------
+
+    /** Test scheduler release when reader is reading buffers. */
+    @Test
+    void testReleasedWhenReading() throws Exception {
+        TestingHsSubpartitionFileReader.Factory factory =
+                new TestingHsSubpartitionFileReader.Factory();
+
+        AtomicReference<Throwable> cause = new AtomicReference<>(null);
+        factory.setFailConsumer((cause::set));
+        factory.setReadBuffersConsumer((buffers, recycle) -> readScheduler.release());
```

Comment:
Depending on the internal implementation, this may cause a deadlock.

Reply:
Rewrite by new Thread to release.